### PR TITLE
add a basic trainer and dataset

### PR DESCRIPTION
**Pull request by manmay-nakhashi**
- huggingface dataset 
- basic trainer

### CODE REVIEW

**Comment by manmay-nakhashi** [comment]
to test or train just run 
by default it's using MushanW/GLOBE dataset 23,519 speakers and covers 164 accents

```
python3 train_e2.py
```

**Comment by lucidrains** [comment]
@manmay-nakhashi Manmay! i remember you now from the natural speech work we did together some time ago

thanks for the PR! I will check it out tomorrow morning 😄

**Comment by lucidrains** [comment]
@manmay-nakhashi hey, looks good! 😄 do you want to try pulling and integrating the text as well?

**Comment by manmay-nakhashi** [comment]
Sure I'll do that.

File: pyproject.toml [hunk 1]
```diff
@@ -30,6 +30,10 @@ dependencies = [
     'torch>=2.0',
     'torchdiffeq',
     'x-transformers>=1.31.12'
+    'accelerate>=0.32.1'
```
Comment (lucidrains) [inline pyproject.toml:33]: missing a comma here

File: e2_trainer.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from torch.nn import functional as F
+from accelerate import Accelerator
+from e2_collate import collate_fn
+import os
+import logging
+from utils.compute_mel import TorchMelSpectrogram
```
Comment (lucidrains) [inline e2_trainer.py:9]: should be `from e2_tts_pytorch.utils.compute_mel`

File: e2_trainer.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from torch.nn import functional as F
+from accelerate import Accelerator
+from e2_collate import collate_fn
```
Comment (lucidrains) [inline e2_trainer.py:6]: should be `e2_tts_pytorch.e2_collate`

File: e2_trainer.py (hunk 2 above)
Comment (manmay-nakhashi) [reply e2_trainer.py:9]: so should i move utils into e2_tts_pytorch/e2_tts_pytorch ?

Comment (lucidrains) [reply e2_trainer.py:9]: @manmay-nakhashi yup, you should move everything under that folder except for `train_e2.py`

**Comment by manmay-nakhashi** [comment]
Screenshot 2024-07-14 at 8 14 32 AM

**Comment by manmay-nakhashi** [comment]
@lucidrains it's ready

**Comment by manmay-nakhashi** [comment]
i'll write a inference script next so we can do some quick experiments.

**Comment by lucidrains** [comment]
nice! it looks good, but in the paper, they didn't use a tokenizer and just went character level

i was thinking we could just use utf character ids?

File: e2_tts_pytorch/utils/compute_mel.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import torch
+from torch import nn
+import torchaudio
```
Comment (lucidrains) [inline e2_tts_pytorch/utils/compute_mel.py:3]: `torchaudio` should be added to the list of dependencies

**Comment by manmay-nakhashi** [comment]
@lucidrains changes are done

File: e2_tts_pytorch/assets/vocab.json [hunk 5]
```diff
@@ -0,0 +1 @@
..."m e","w as","g h","i d","l l","w i","en t","f or","a y","r o","v er","i c","h er","k e","h is","n o","u t","u n","i r","l o","w e","r i","h a","wi th","gh t","ou t","i m","i on","al l","a b","on e","n e","g e","ou ld","t er","m o","h ad","c e","s he","g o","s h","u r","a m","s o","p e","m y","d e","a re","b ut","om e","f r","the r","f e","s u","d o","c on","t e","a in","er e","p o","i f","the y","u s","a g","t r","n ow","ou n","th is","ha ve","no t","s a","i l","u p","th ing","fr om","a p","h im","ac k","at ion","an t","ou r","o p","li ke","u st","es s","b o","o k","u l","in d","e x","c om","s ome","the re","er s","c o","re s","m an","ar d","p l","w or","w ay","ti on","f o","c a","w ere","b y","at e","p ro","t ed","oun d","ow n","w ould","t s","wh at","q u","al ly","i ght","c k","g r","wh en","v en","c an","ou gh","in e","en d","p er","ou s","o d","id e","k now","t y","ver y","s i","a k","wh o","ab out","i ll","the m","es t","re d","y e","c ould","on g","you r","the ir","e m","j ust","o ther","in to","an y","wh i","u m","t w","as t","d er","d id","i e","be en","ac e","in k","it y","b ack","t ing","b r","mo re","a ke","p p","the n","s p","e l","u se","b l","sa id","o ver","ge t"]}}
```
Comment (lucidrains) [inline e2_tts_pytorch/assets/vocab.json:1]: this file can be removed now

File: train_e2.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import torch
+import torch.nn.init as init
```
Comment (lucidrains) [inline train_e2.py:2]: think `init` is never used

File: e2_tts_pytorch/utils/compute_mel.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        hop_length=256,
+        win_length=1024,
+        n_mel_channels=80,
+        mel_fmin=0,
+        mel_fmax=8000,
+        sampling_rate=22050,
+        normalize=False,
+    ):
+        super().__init__()
+        self.filter_length = filter_length
```
Comment (lucidrains) [inline e2_tts_pytorch/utils/compute_mel.py:17]: lines `16-23` can be omitted, as most of the arguments are forwarded to `MelSpectrogram`

File: e2_tts_pytorch/dataset/e2_dataset.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+logger = logging.getLogger(__name__)
+
+class E2EDataset(Dataset):
+    def __init__(self, hf_dataset):
+        self.data = load_dataset(hf_dataset, split='train')
+        self.target_sample_rate = 22050
+        self.hop_length = 256
+        self.mel_spectrogram = TorchMelSpectrogram(sampling_rate=self.target_sample_rate)
+        
+        self.char_set = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,!?-:;'\"()[] ")
```
Comment (lucidrains) [inline e2_tts_pytorch/dataset/e2_dataset.py:18]: do you want to try just using https://github.com/lucidrains/e2-tts-pytorch/blob/main/e2_tts_pytorch/e2_tts.py#L55 ?

**Comment by manmay-nakhashi** [comment]
resolve all the suggestions

**Comment by lucidrains** [comment]
@manmay-nakhashi thank you Manmay!

**Pull request merged**